File: src/leetcode/2026/02/Solution566.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution566 {
public:
    vector<vector<int>> matrixReshape(vector<vector<int>>& mat, int r, int c) {
        int originalRows = mat.size();
        int originalCols = mat[0].size();
        
        if (originalRows * originalCols != r * c) {
            return mat; // Return original matrix if reshape is not possible
        }
        
        vector<vector<int>> reshapedMatrix(r, vector<int>(c));
        for (int i = 0; i < originalRows * originalCols; ++i) {
            reshapedMatrix[i / c][i % c] = mat[i / originalCols][i % originalCols];
        }
        
        return reshapedMatrix;
    }
};

int main() {
    // Example usage:
    Solution566 solution;

    vector<vector<int>> mat1 = {{1, 2}, {3, 4}};
    int r1 = 1, c1 = 4;
    vector<vector<int>> reshaped = solution.matrixReshape(mat1, r1, c1);
    // reshaped should be {{1, 2, 3, 4}}
    for (vector<int> row : reshaped) {
        for (int val : row) {
            cout << val << " ";
        }
        cout << endl;
    }

    vector<vector<int>> mat2 = {{1, 2}, {3, 4}};
    int r2 = 2, c2 = 4;
    reshaped = solution.matrixReshape(mat2, r2, c2);
    // reshaped should be {{1, 2}, {3, 4}} (reshape not possible)
    for (vector<int> row : reshaped) {
        for (int val : row) {
            cout << val << " ";
        }
        cout << endl;
    }
    return 0;
}

File: src/leetcode/2026/02/Solution566.java
public class Solution566 {
    public int[][] matrixReshape(int[][] mat, int r, int c) {
        int originalRows = mat.length;
        int originalCols = mat[0].length;

        if (originalRows * originalCols != r * c) {
            return mat;
        }

        int[][] reshapedMatrix = new int[r][c];
        for (int i = 0; i < originalRows * originalCols; i++) {
            reshapedMatrix[i / c][i % c] = mat[i / originalCols][i % originalCols];
        }

        return reshapedMatrix;
    }

    public static void main(String[] args) {
        Solution566 solution = new Solution566();

        // Example 1
        int[][] mat1 = { { 1, 2 }, { 3, 4 } };
        int r1 = 1;
        int c1 = 4;
        int[][] output1 = solution.matrixReshape(mat1, r1, c1);
        for (int[] row : output1) {
            for (int val : row) {
                System.out.print(val + " ");
            }
            System.out.println();
        }
        // Output:
        // 1 2 3 4

        // Example 2
        int[][] mat2 = { { 1, 2 }, { 3, 4 } };
        int r2 = 2;
        int c2 = 4;
        int[][] output2 = solution.matrixReshape(mat2, r2, c2);
        for (int[] row : output2) {
            for (int val : row) {
                System.out.print(val + " ");
            }
            System.out.println();
        }
        // Output:
        // 1 2
        // 3 4
    }
}
